hugr-model/src/v0/scope/symbol.rs

```diff
@@ -0,0 +1,208 @@
+                debug_assert_eq!(last.unwrap().1, self.bindings.len());
+            }
+        }
+    }
+
+    /// Insert a new symbol into the current scope.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the symbol is already defined in the current scope.
+    /// In the case of an error the table remains unchanged.
+    ///
+    /// # Panics
+    ///
+    /// Panics if there is no current scope.
+    pub fn insert(&mut self, name: &'a str, node: NodeId) -> Result<(), DuplicateSymbolError> {
+        let scope_depth = self.scopes.len() as u16 - 1;
+        let (symbol_index, shadowed) = self.symbols.insert_full(name, self.bindings.len());
+
+        if let Some(shadowed) = shadowed {
+            let (shadowed_node, shadowed_binding) = self.bindings.get_index(shadowed).unwrap();
+            if shadowed_binding.scope_depth == scope_depth {
+                self.symbols.insert(name, shadowed);
+                return Err(DuplicateSymbolError(name.into(), node, *shadowed_node));
+            }
+        }
+
+        self.bindings.insert(
+            node,
+            Binding {
+                scope_depth,
+                shadows: shadowed,
+                symbol_index,
+            },
+        );
+
+        Ok(())
+    }
+
+    /// Check whether a symbol is currently visible in the current scope.
+    pub fn is_visible(&self, node: NodeId) -> bool {
+        let Some(binding) = self.bindings.get(&node) else {
+            return false;
+        };
+
+        // Check that the symbol has not been shadowed at this point.
+        self.symbols[binding.symbol_index] == binding.symbol_index
+    }
+
+    /// Tries to resolve a symbol name in the current scope.
+    pub fn resolve(&self, name: &'a str) -> Result<NodeId, UnknownSymbolError> {
+        let index = *self
+            .symbols
+            .get(name)
+            .ok_or(UnknownSymbolError(name.into()))?;
+
+        // NOTE: The unwrap is safe because the `symbols` map
+        // points to valid indices in the `bindings` map.
+        let (node, _) = self.bindings.get_index(index).unwrap();
+        Ok(*node)
+    }
+
+    /// Returns the depth of the given region, if it corresponds to a currently open scope.
+    pub fn region_to_depth(&self, region: RegionId) -> Option<ScopeDepth> {
+        Some(self.scopes.get_index_of(&region)? as _)
+    }
+
+    /// Returns the region corresponding to the scope at the given depth.
+    pub fn depth_to_region(&self, depth: ScopeDepth) -> Option<RegionId> {
+        let (region, _) = self.scopes.get_index(depth as _)?;
+        Some(*region)
+    }
+
+    /// Resets the symbol table to its initial state while maintaining its
+    /// allocated memory.
+    pub fn clear(&mut self) {
+        self.symbols.clear();
+        self.bindings.clear();
+        self.scopes.clear();
+    }
+}
+
+impl Default for SymbolTable<'_> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+struct Binding {
```

Comment:
If you put these `struct`s up after the `struct SymbolTable` but before `impl SymbolTable` I think it'd make this much easier to read, in that you'd see things in a order useful for comprehension